README.md
[![Travis build status](https://travis-ci.org/arnaldpuy/sensobol.svg?branch=master)](https://travis-ci.org/arnaldpuy/sensobol) [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/arnaldpuy/sensobol?branch=master&svg=true)](https://ci.appveyor.com/project/arnaldpuy/sensobol) [![CRAN_Status_Badge_version_last_release](https://www.r-pkg.org/badges/version-last-release/sensobol)](https://cran.r-project.org/package=sensobol) [![metacran downloads](https://cranlogs.r-pkg.org/badges/last-week/sensobol)](https://cran.r-project.org/package=sensobol) [![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/sensobol)](https://cran.r-project.org/package=sensobol) [![Coverage status](https://codecov.io/gh/arnaldpuy/sensobol/branch/master/graph/badge.svg)](https://codecov.io/github/arnaldpuy/sensobol?branch=master) [![license](https://img.shields.io/badge/license-GPL--3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2579855.svg)](https://doi.org/10.5281/zenodo.2579855)

# sensobol

The goal of `sensobol` is to provide a set of functions to swiftly compute and visualize up to third-order Sobol' sensitivity indices. The functions allow to: 
- Create the sample matrices for the model evaluation.
- Compute and bootstrap up to third-order effects.
- Assess the approximation error of Sobol' indices.
- Plot the model uncertainty and the Sobol' indices.

## Installation
To install the stable version on [CRAN](https://CRAN.R-project.org/package=sensobol), use

```r
install.packages("sensobol")
```
To install the development version, use devtools:

``` r
install.packages("devtools") # if you have not installed devtools package already
devtools::install_github("arnaldpuy/sensobol", build_vignettes = TRUE)
```

## Example

This brief example shows how to compute Sobol' indices. For a more detailed explanation of the package functions, check the vignette.

``` r
## Load the package:
library(sensobol)

## Create sample matrix to compute first, total and second-order indices:
A <- sobol_matrices(n = 1000, k = 3,  second = TRUE)

## Compute the model output (using the Ishigami test function):
Y <- ishigami_Mapply(A)

## Compute the Sobol' indices (first, total and second-order):
sens <- sobol_indices(Y = Y, params = colnames(data.frame(A)), R = 100, n = 1000, second = TRUE)
```

## Citation

Please use the following citation if you use `sensobol` in your publications:

```r
Arnald Puy (2019). sensobol: Computation of High-Order Sobol' Sensitivity Indices. R package
  version 0.2.0 http://github.com/arnaldpuy/sensobol
```

A BibTex entry for LaTex users is:

```r
@Manual{,
    title = {sensobol: Computation of High-Order Sobol' Sensitivity Indices},
    author = {Arnald Puy},
    year = {2019},
    note = {R package version 0.2.0},
    url = {http://github.com/arnaldpuy/sensobol},
  }
```


NEWS.md
# sensobol 1.1.2

* Fixed a bug on sobol_matrices that caused the function to fail when a single parameter was passed into the "params" argument.

* Added a line in the help page of the sobol_indices and sobol_dummy functions informing that the Y argument does not accept NA or NaN values in the model output.

# sensobol 1.1.1

* The CITATION file has changed. Citations to sensobol should now refer to the paper published in the Journal of Statistical Software: 

Arnald Puy, Samuele Lo Piano, Andrea Saltelli, and Simon A. Levin. sensobol: an R package to compute variance-based sensitivity indices. Journal of Statistical Software 102.5 (2022), pp. 1-37. doi: 10.18637/jss.v102.i05

* The metafunction has been extended with a piecewise large, piecewise small and an oscillation function.

# sensobol 1.1.0

* The package now includes support to calculate and plot up to fourth-order effects.

# sensobol 1.0.4

* Fixed redundant links in the bibliography in .Rd files.

# sensobol 1.0.3

* Added a help page for the package.

* Corrected a bug in sobol_ode.

# sensobol 1.0.2

* The function `plot_sensobol`has been eliminated. Now the Sobol' indices can be plot by a single call to `plot`.

* Third-order indices are now referred to as `Sijl` rather than `Sijk`.

* The function `sobol_convergence`has been added to check the convergence of
Sobol' indices across sub-samples of the model output.

# sensobol 1.0.1

* The output of `sobol_indices` is now an object of class `sensobol`. Besides including the indices, it also informs on the sum of first-order effects, the estimators used in the computation and the total number of model runs.

* The output of `vars_to` is now an object of class `vars`. Besides including the indices, it also informs on the number of stars and the h value used.

* The function `plot_sobol` is deprecated and will be removed from future versions. Now the output of `sobol_indices`can be printed with a call to `plot`.

# sensobol 1.0.0

* This is a major package upgrade.

* The package now includes four first-order (Sobol', Salteli, Jansen and Azzini) and eight total-order (Jansen, Sobol', Homma, Saltelli, Janon, Glen, Azzini and VARS-TO) sensitivity estimators.

* The sample matrix can be constructed either with Sobol' quasi-random numbers, a latin hypercube design or random numbers.

* Several functions to plot the results of the uncertainty and sensitivity analysis have been incorporated: the function `plot_scatter` plots the model inputs against the output, whereas the function `plot_multiscatter` plots x_i against x_j and maps the resulting coordinate to its respective model output value.

* The package can now be used in models with either a scalar or a multivariate output.

# sensobol 0.2.2

* This release prepares the users for a major upcoming improvement of the package.

* Added warnings to the function `sobol_matrices`. The arguments `n` and `k` will
be substituted by `N` and `params` in the next release of the package. The 
arguments `second` and `third` will be substituted by the argument `order` in the next
release of the package.

* Added warnings to the functions `plot_uncertainty`, `plot_scatter`, `sobol_dummy` and `sobol_indices`. The argument `n` will be substituted by `N` in the next release of the package. 

* The function `sobol_ci` is deprecated. The computation of confidence intervals
will be done directly by `sobol_indices` in the next version of the package.

* The function `sobol_ci_dummy` is deprecated. The computation of confidence intervals
for dummy parameters will be done directly by `sobol_dummy` in the next version of the package.

* The function `sobol_replicas` is deprecated and will be removed in the next version
of the package.

* The test function `ishigami_Mapply` would be renamed as `ishigami_Fun` in the next
release of the package.

# sensobol 0.2.1

* Corrected a bug in the `plot_uncertainty` function. Now the function
demands to add the initial sample size `n` to visualize the model output uncertainty.

* Corrected some functions to adapt them to data.table 1.12.4.

* Added a new function (`sobol_replicas`) to easily extract the bootstrapped Sobol' indices.

# sensobol 0.2.0

* New option in the `sobol_matrices` function: the option `cluster` allows to create Sobol' matrices for clusters of parameters.

* New test functions added: 
  - Bratley & Fox 1988.
  - Bratley et al. 1992
  - Oakley & O'Hagan 2004
  
* Added references to all test functions.

# sensobol 0.1.1

* The vignette rendered wrongly in the previous version;
now the issue is corrected.

* Corrected the following note, found in the CRAN check: 
  - Namespace in Imports field not imported from: ‘parallel’.
 
* Some functions that were exported to the R package manual 
are now internal.

# sensobol 0.1.0

* Added a NEWS.md file to track changes to the package.



R/test_functions.R

# Sobol ' G function -----------------------------------------------------------

#' Sobol' G function
#'
#' It implements the \insertCite{Sobol1998;textual}{sensobol} function. In this
#' case, the function works with 8 model inputs.
#'
#' @param X A data frame or numeric matrix.
#'
#' @return A numeric vector with the model output.
#' @export
#'
#' @references
#' \insertAllCited{}
#' @examples
#' A <- sobol_matrices(n = 100, k = 8)
#' Y <- sobol_Fun(A)
sobol_Fun <- function(X) {
  a <- c(0, 1, 4.5, 9, 99, 99, 99, 99)
  y <- 1
  for (j in 1:8) {
    y <- y * (abs(4 * X[, j] - 2) + a[j])/(1 + a[j])
  }
  return(y)
}

# Ishigami function -----------------------------------------------------------

ishigami <- function(X1, X2, X3) {
  A <- 2
  B <- 1
  sin(X1) + A * sin(X2) ^ 2 + B * X3 ^ 4 * sin(X1)
}

#' Ishigami function
#'
#' It implements the \insertCite{Ishigami1990;textual}{sensobol} function,
#' which requires 3 model inputs. The transformation of the
#' distribution of the model inputs (from U(0, 1) to U(-pi, +pi)) is conducted
#' internally.
#'
#' @param X A data frame or numeric matrix.
#'
#' @return A numeric vector with the model output.
#' @export
#'
#' @references
#' \insertAllCited{}
#'
#' @examples
#' A <- sobol_matrices(n = 100, k = 3)
#' Y <- ishigami_Mapply(A)
ishigami_Mapply <- function(X) {
  warning("ishigami_Mapply will be substituted by ishigami_Fun in the next release of the package")
  X <- apply(X, 2, function(x) x * (pi + pi) - pi)
  return(mapply(ishigami,
                X[, 1],
                X[, 2],
                X[, 3]))
}

# Bratley et al. function -----------------------------------------------------


#' Bratley, Fox and Niederreiter (1992) function
#'
#' It implements the \insertCite{Bratley1992;textual}{sensobol} function,
#' #' which requires \emph{n} model inputs.
#'
#' @param X A data frame or numeric matrix.
#'
#' @return A numeric vector with the model output.
#' @export
#' @references
#' \insertAllCited{}
#'
#' @examples
#' A <- sobol_matrices(n = 100, k = 4)
#' Y <- bratley1992_Fun(A)
bratley1992_Fun <- function(X) {
  # Preallocate
  mat <- tmp <- vector(mode = "list", length = nrow(X))
  Y <- vector(mode = "numeric", length = nrow(X))
  for(i in 1:nrow(X)) {
    mat[[i]] <- matrix(rep(X[i, ], times = ncol(X)),
                       nrow = ncol(X),
                       ncol = ncol(X),
                       byrow = TRUE)
    mat[[i]][upper.tri(mat[[i]])] <- 1
    tmp[[i]] <- matrixStats::rowProds(mat[[i]])
    Y[[i]] <- sum(tmp[[i]] * (-1) ^ (1:ncol(X)))
  }
  return(Y)
}

# Bratley and Fox (1988) function ---------------------------------------------

#' Bratley and Fox (1988) function
#'
#' It implements the \insertCite{Bratley1988;textual}{sensobol} function,
#' which requires \emph{n} model inputs.
#'
#' @param X A data frame or numeric matrix.
#'
#' @return A numeric vector with the model output.
#'
#' @export
#'
#' @examples
#' A <- sobol_matrices(n = 100, k = 6)
#' Y <- bratley1988_Fun(A)
bratley1988_Fun <- function(X) {
  y <- 1
  for (j in 1:ncol(X)) {
    y <- y * (abs(4 * X[, j] - 2))}
  return(y)
}

# Oakley and O'Hagan function -------------------------------------------------

#' Oakley & O'Hagan (2004) function
#'
#' It implements the \insertCite{Oakley2004;textual}{sensobol} function.
#' The function needs 15 model inputs. The transformation of the
#' distribution of the model inputs (from uniform to normally distributed
#' with mean = 0 and sd = 1) is conducted internally.
#'
#' @param X A data frame or numeric matrix.
#'
#' @return A numeric vector with the model output.
#'
#' @export
#' @references
#' \insertAllCited{}
#'
#' @examples
#' A <- sobol_matrices(n = 100, k = 15)
#' Y <- oakley_Fun(A)
oakley_Fun <- function(X) {

  a1 <- c(0.0118, 0.0456, 0.2297, 0.0393, 0.1177,
          0.3865, 0.3897, 0.6061, 0.6159, 0.4005,
          1.0741, 1.1474, 0.788, 1.1242, 1.1982)

  a2 <- c(0.4341, 0.0887, 0.0512, 0.3233, 0.1489,
          1.036, 0.9892, 0.9672, 0.8977, 0.8083,
          1.8426, 2.4712, 2.3946, 2.0045, 2.2621)

  a3 <- c(0.1044, 0.2057, 0.0774, 0.273, 0.1253,
          0.7526, 0.857, 1.0331, 0.8388, 0.797,
          2.2145, 2.0382, 2.4004, 2.0541, 1.9845)

  M <- c(-0.022482886, -0.18501666, 0.13418263, 0.36867264, 0.17172785, 0.13651143, -0.44034404,
         -0.081422854, 0.71321025, -0.44361072, 0.50383394, -0.024101458, -0.045939684, 0.21666181,
         0.055887417, 0.2565963, 0.053792287, 0.25800381, 0.23795905, -0.59125756, -0.081627077,
         -0.28749073, 0.41581639, 0.49752241, 0.083893165, -0.11056683, 0.033222351, -0.13979497,
         -0.031020556, -0.22318721, -0.055999811, 0.19542252, 0.095529005, -0.2862653, -0.14441303,
         0.22369356, 0.14527412, 0.28998481, 0.2310501, -0.31929879, -0.29039128, -0.20956898, 0.43139047,
         0.024429152, 0.044904409, 0.66448103, 0.43069872, 0.29924645, -0.16202441, -0.31479544,
         -0.39026802, 0.17679822, 0.057952663, 0.17230342, 0.13466011, -0.3527524, 0.25146896, -0.018810529,
         0.36482392, -0.32504618, -0.121278, 0.12463327, 0.10656519, 0.046562296, -0.21678617, 0.19492172,
         -0.065521126, 0.024404669, -0.09682886, 0.19366196, 0.33354757, 0.31295994, -0.083615456,
         -0.25342082, 0.37325717, -0.2837623, -0.32820154, -0.10496068, -0.22073452, -0.13708154,
         -0.14426375, -0.11503319, 0.22424151, -0.030395022, -0.51505615, 0.017254978, 0.038957118,
         0.36069184, 0.30902452, 0.050030193, -0.077875893, 0.003745656, 0.88685604, -0.26590028,
         -0.079325357, -0.042734919, -0.18653782, -0.35604718, -0.17497421, 0.088699956, 0.40025886,
         -0.055979693, 0.13724479, 0.21485613, -0.011265799, -0.09229473, 0.59209563, 0.031338285,
         -0.033080861, -0.24308858, -0.099798547, 0.034460195, 0.095119813, -0.3380162, 0.0063860024,
         -0.61207299, 0.081325416, 0.88683114, 0.14254905, 0.14776204, -0.13189434, 0.52878496, 0.12652391,
         0.045113625, 0.58373514, 0.37291503, 0.11395325, -0.29479222, -0.57014085, 0.46291592, -0.094050179,
         0.13959097, -0.38607402, -0.4489706, -0.14602419, 0.058107658, -0.32289338, 0.093139162,
         0.072427234, -0.56919401, 0.52554237, 0.23656926, -0.011782016, 0.071820601, 0.078277291,
         -0.13355752, 0.22722721, 0.14369455, -0.45198935, -0.55574794, 0.66145875, 0.34633299, 0.14098019,
         0.51882591, -0.28019898, -0.1603226, -0.068413337, -0.20428242, 0.069672173, 0.23112577,
         -0.044368579, -0.16455425, 0.21620977, 0.0042702105, -0.087399014, 0.31599556, -0.027551859,
         0.13434254, 0.13497371, 0.05400568, -0.17374789, 0.17525393, 0.060258929, -0.17914162, -0.31056619,
         -0.25358691, 0.025847535, -0.43006001, -0.62266361, -0.033996882, -0.29038151, 0.03410127,
         0.034903413, -0.12121764, 0.026030714, -0.33546274, -0.41424111, 0.05324838, -0.27099455,
         -0.026251302, 0.41024137, 0.26636349, 0.15582891, -0.18666254, 0.019895831, -0.24388652,
         -0.44098852, 0.012618825, 0.24945112, 0.071101888, 0.24623792, 0.17484502, 0.0085286769,
         0.2514707, -0.14659862, -0.08462515, 0.36931333, -0.29955293, 0.1104436, -0.75690139, 0.041494323,
         -0.25980564, 0.46402128, -0.36112127, -0.94980789, -0.16504063, 0.0030943325, 0.052792942,
         0.22523648, 0.38390366, 0.45562427, -0.18631744, 0.0082333995, 0.16670803, 0.16045688)

  M <- matrix(M, 15, 15, byrow = TRUE)

  Y <- vector()
  # transformation to normal distribution
  X <- apply(X, 2, function(x) stats::qnorm(x))
  for(i in 1:nrow(X)) {
    mat <- matrix(X[i, ])
    Y[i] <- a1 %*% mat + a2 %*% sin(mat) + a3 %*% cos(mat) + t(mat) %*% M %*% mat
  }
  return(Y)
}


R/plot_sobol.R


# PLOT SOBOL' FIRST AND TOTAL-ORDER INDICES -----------------------------------

#' Plot Sobol' first and total-order indices
#'
#' @param x A data.table.
#' @param dummy The output of the \code{sobol_ci_dummy} function. If supplied and
#' \code{type = 1}, the plot includes an horizontal transparent frame showing the confidence
#' intervals of the first and total-order indices for the dummy parameter.
#' @param type An integer. If \code{type = 1}, it plots first and total effects.
#' If \code{type = 2}, it plots second-order effects. If \code{type = 3}, it plots
#' third-order effects. Default is \code{type = 1}.
#'
#' @return A ggplot object.
#' @import ggplot2
#' @export
#'
#' @examples
#' # Define settings:
#' n <- 500; k <- 8; R <- 100
#' # Design the sample matrix:
#' A <- sobol_matrices(n = n, k = k, second = TRUE, third = TRUE)
#' # Compute the model output:
#' Y <- sobol_Fun(A)
#' # Compute the Sobol' indices:
#' \donttest{sens <- sobol_indices(Y = Y, params = colnames(data.frame(A)),
#' R = R, n = n, parallel = "no", ncpus = 1, second = TRUE, third = TRUE)
#' # Compute the Sobol' indices for the dummy parameter:
#' s.dummy <- sobol_dummy(Y = Y, params = colnames(data.frame(A)), R = R, n = n)
#' # Compute confidence intervals:
#' sens.ci <- sobol_ci(sens, params = colnames(data.frame(A)), type = "norm", conf = 0.95)
#' # Compute confidence intervals for the dummy parameter:
#' s.dummy.ci <- sobol_ci_dummy(s.dummy, type = "norm", conf = 0.95)
#' # Plot Sobol' indices:
#' plot_sobol(sens.ci, dummy = s.dummy.ci, type = 1)}
plot_sobol <- function(x, dummy = NULL, type = 1) {
  sensitivity <- low.ci <- high.ci <- parameters <- original <- NULL
  if(type == 1) {
    if(is.null(dummy) == FALSE) {
      plot.dummy <- geom_rect(data = dummy,
                              aes(ymin = 0,
                                  ymax = high.ci,
                                  xmin = -Inf,
                                  xmax = Inf,
                                  fill = sensitivity),
                              alpha = 0.2,
                              inherit.aes = FALSE)
    } else {
      plot.dummy <- NULL
    }
    p <- x[sensitivity == "Si" | sensitivity == "STi"]
    gg <- ggplot2::ggplot(p, aes(parameters, original,
                                 fill = sensitivity)) +
      geom_bar(stat = "identity",
               position = position_dodge(0.6),
               color = "black") +
      plot.dummy +
      geom_errorbar(aes(ymin = low.ci,
                        ymax = high.ci),
                    position = position_dodge(0.6)) +
      scale_fill_discrete(name = "Sobol' indices",
                          labels = c(expression(S[italic(i)]),
                                     expression(S[italic(T[i])]))) +
      labs(x = "",
           y = "Variance") +
      theme_bw() +
      theme(legend.position = "top",
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            legend.background = element_rect(fill = "transparent",
                                             color = NA),
            legend.key = element_rect(fill = "transparent",
                                      color = NA))
  } else if(!type == 1) {
    if(type == 2) {
      plot.type <- "Sij"
    } else if(type == 3) {
      plot.type <- "Sijk"
    } else {
      stop("Type should be either 1, 2 or 3")
    }
    p <- x[sensitivity == plot.type]
    gg <- ggplot2::ggplot(p, aes(stats::reorder(parameters, original),
                    original)) +
      geom_point() +
      geom_errorbar(aes(ymin = low.ci,
                        ymax = high.ci)) +
      theme_bw() +
      labs(x = "",
           y = "Variance") +
      geom_hline(yintercept = 0,
                 lty = 2,
                 color = "red") +
      theme(panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            legend.background = element_rect(fill = "transparent",
                                             color = NA),
            legend.key = element_rect(fill = "transparent",
                                      color = NA),
            axis.text.x = element_text(angle = 45,
                                       hjust = 1))
  }
  return(gg)
}

# PLOT MODEL OUTPUT UNCERTAINTY -----------------------------------------------

#' Plot model output uncertainty
#'
#' It creates an histogram with the model output distribution.
#'
#' @param Y A numeric vector with the model output.
#' @param n An integer with the initial sample size.
#'
#' @return a ggplot2 object.
#' @import ggplot2
#' @export
#'
#' @examples
#' # Define settings:
#' n <- 100; k <- 8; R <- 10
#' # Design the sample matrix:
#' A <- sobol_matrices(n = n, k = k, second = FALSE, third = FALSE)
#' # Compute the model output:
#' Y <- sobol_Fun(A)
#' # Plot the model output distribution:
#' plot_uncertainty(Y, n = n)
plot_uncertainty <- function(Y, n = NULL) {
  warning("The argument n will be substituted by N in the next release of the package")
  if(is.vector(Y) == FALSE) {
    stop("Y should be a vector")
  }
  if(is.null(n) == TRUE) {
    stop("n should be specified")
  }
  Y <- Y[1:(2*n)]
  df <- data.frame(Y)
  gg <- ggplot2::ggplot(df, aes(Y)) +
    geom_histogram(color = "black",
                   fill = "white") +
    labs(x = "Y",
         y = "Count") +
    theme_bw() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          legend.background = element_rect(fill = "transparent",
                                           color = NA),
          legend.key = element_rect(fill = "transparent",
                                    color = NA))
  return(gg)
}

# PLOT SCATTERPLOT OF MODEL OUTPUT VERSUS MODEL INPUTS ------------------------

#' Scatterplots of the model output against the model inputs
#'
#' @param x A data table, data frame or matrix with the
#' model inputs.
#' @param Y Numeric vector with the model output.
#' @param n Integer, sample size of the Sobol' matrix.
#' @param params Vector with the name of the model inputs.
#'
#' @return A ggplot object.
#' @import ggplot2
#' @export
#'
#' @examples
#' # Define settings:
#' n <- 100; k <- 8; R <- 10
#' # Design the sample matrix:
#' A <- sobol_matrices(n = n, k = k, second = TRUE, third = TRUE)
#' # Compute the model output:
#' Y <- sobol_Fun(A)
#' # Plot scatterplots:
#' plot_scatter(x = A, Y = Y, n = n, params = colnames(data.frame(A)))
plot_scatter <- function(x, Y, n, params) {
  warning("The argument n will be substituted by N in the next release of the package")
  value <- NULL
  dt <- data.table::data.table(cbind(x, Y))
  # Retrieve the A and B matrices only
  dt <- dt[1:(2 * n)]
  data.table::setnames(dt, 1:ncol(dt), paste(c(params, "Y")))
  dt.melted <- data.table::melt(dt, measure.vars = 1:length(params),
                                variable.name = "Parameters")
  gg <- ggplot2::ggplot(dt.melted, aes(x = value, y = Y)) +
    geom_hex() +
    labs(x = "Value",
         y = "Y") +
    facet_wrap(~Parameters,
               scales = "free_x") +
    theme_bw() +
    scale_x_continuous(breaks = scales::pretty_breaks(n = 3)) +
    scale_fill_continuous(name = "Count") +
    theme(legend.position = "top",
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          legend.background = element_rect(fill = "transparent",
                                           color = NA),
          legend.key = element_rect(fill = "transparent",
                                    color = NA))
  return(gg)
}


R/sobol_boot.R


# EXTRACT BOOTSTRAP REPLICAS --------------------------------------------------

#' Extract bootstrap replicas
#'
#' It creates a `data.table` with all the bootstrap replicas.
#'
#' @param dt The output of the `sobol_indices` function.
#' @param k The number of parameters.
#' @param second Logical. If \code{second = TRUE}, it computes the confidence
#' intervals for second-order indices. Default is \code{second = FALSE}.
#' @param third Logical. If \code{third = TRUE}, it computes the confidence
#' intervals for third-order indices. Default is \code{third = FALSE}.
#'
#' @importFrom utils "combn"
#' @importFrom rlang ":="
#' @importFrom data.table ".SD"
#'
#' @return A `data.table`.
#' @export
#'
#' @examples
#' # Define settings:
#' n <- 100; k <- 8; R <- 100
#' # Design the sample matrix:
#' A <- sobol_matrices(n = n, k = k, second = TRUE, third = TRUE)
#' # Compute the model output:
#' Y <- sobol_Fun(A)
#' # Compute the Sobol' indices:
#' \donttest{sens <- sobol_indices(Y = Y, params = colnames(data.frame(A)),
#' R = R, n = n, parallel = "no", ncpus = 1, second = TRUE, third = TRUE)
#' # Extract the bootstrap replicas up to third-order
#' replicas <- sobol_replicas(dt = sens, k = k, second = TRUE, third = TRUE)}

sobol_replicas <- function(dt, k, second = FALSE, third = FALSE) {
  .Deprecated(msg = "sobol_replicas will be removed in the next version")
  All <- V1 <- parameters <- replicas <- NULL
  # Extract first and total-order replicas
  first.and.total <- dt[1:k, "All":= list(lapply(V1, function(x) x["t"]))][
    , list("Si" = lapply(All, function(x) lapply(x, function(y) y[, 1])),
           "STi" = lapply(All, function(x) lapply(x, function(y) y[, 2]))),
      parameters][
    , lapply(.SD, unlist), .SDcols = c("Si", "STi"), parameters]
  first.and.total <- data.table::melt(first.and.total, measure.vars = c("Si", "STi"))
  out <- first.and.total
  if(second == TRUE) {
    # Extract up to second-order replicas
    second <- dt[(k + 1): (k + length(combn(1:k, 2, simplify = FALSE)))][
      , "replicas":= list(lapply(V1, function(x) x["t"]))][
      , list("Sij" = lapply(replicas, function(x) lapply(x, function(y) y[, 1]))), parameters][
      , lapply(.SD, unlist), .SDcols = "Sij", parameters]
    second <- data.table::melt(second, measure.vars = "Sij")
    out <- rbind(first.and.total, second)
  }
  if(third == TRUE) {
    # Extract up to third-order replicas
    first.plus.second <- k + (length(combn(1:k, 2, simplify = FALSE)))
    third <- dt[(first.plus.second + 1):(first.plus.second + length(combn(1:k, 3, simplify = FALSE)))][
      , "replicas":= list(lapply(V1, function(x) x["t"]))][
      , list("Sijk" = lapply(replicas, function(x) lapply(x, function(y) y[, 1]))), parameters][
      , lapply(.SD, unlist), .SDcols = "Sijk", parameters]
    third <- data.table::melt(third, measure.vars = "Sijk", variable.name = "sensitivity")
    out <- rbind(first.and.total, second, third)
  }
  return(out)
}

# FUNCTION TO COMPUTE BOOTSTRAP CONFIDENCE INTERVALS --------------------------

bootstats <- function(b, conf = conf, type = type) {
  p <- length(b$t0)
  lab <- c("original", "bias", "std.error", "low.ci", "high.ci")
  tmp <- as.data.frame(matrix(nrow = p,
                              ncol = length(lab),
                              dimnames = list(NULL, lab)))
  for (i in 1:p) {
    # original estimation, bias, standard deviation
    tmp[i, "original"] <- b$t0[i]
    tmp[i, "bias"] <- mean(b$t[, i]) - b$t0[i]
    tmp[i, "std.error"] <- stats::sd(b$t[, i])
    # confidence interval
    if (type == "norm") {
      ci <- boot::boot.ci(b, index = i, type = "norm", conf = conf)
      if (!is.null(ci)) {
        tmp[i, "low.ci"] <- ci$norm[2]
        tmp[i, "high.ci"] <- ci$norm[3]
      }
    } else if (type == "basic") {
      ci <- boot::boot.ci(b, index = i, type = "basic", conf = conf)
      if (!is.null(ci)) {
        tmp[i, "low.ci"] <- ci$basic[4]
        tmp[i, "high.ci"] <- ci$basic[5]
      }
    } else if (type == "percent") {
      ci <- boot::boot.ci(b, index = i, type = "perc", conf = conf)
      if (!is.null(ci)) {
        tmp[i, "low.ci"] <- ci$percent[4]
        tmp[i, "high.ci"] <- ci$percent[5]
      }
    } else if (type == "bca") {
      ci <- boot::boot.ci(b, index = i, conf = conf)
      if (!is.null(ci)) {
        tmp[i, "low.ci"] <- ci$bca[4]
        tmp[i, "high.ci"] <- ci$bca[5]
      }
    }
  }
  return(tmp)
}

sobol_ci_temp <- function(b, params, type, conf, second = FALSE, third = FALSE) {
  V1 <- NULL
  # Extract first and total order effects
  ext1 <- b[, .SD[1:length(params)]][, V1]
  Si.STi <- lapply(ext1, function(x) bootstats(x,
                                               type = type,
                                               conf = conf)) %>%
    data.table::rbindlist()
  if(second == TRUE) {
    # Extract second order effects
     ext2 <- b[, .SD[(length(params) + 1) :(length(params) +
                                              factorial(length(params)) /
                                              (factorial(2) * factorial(length(params) - 2)))]][, V1]
     Sij <- lapply(ext2, function(x) bootstats(x,
                                               type = type,
                                               conf = conf)) %>%
       data.table::rbindlist()
  } else {
    Sij <- NULL
  }
  if(third == TRUE) {
    ext3 <- b[, utils::tail(.SD, factorial(length(params)) /
                              (factorial(3) * factorial(length(params) - 3)))][, V1]
    Sijk <- lapply(ext3, function(x) bootstats(x,
                                               type = type,
                                               conf = conf)) %>%
      data.table::rbindlist()
  } else {
    Sijk <- NULL
  }
  return(rbind(Si.STi, Sij, Sijk))
}

# CREATE TWO VECTORS WITH SENSITIVITY INDICES AND MODEL INPUTS ----------------

create_vectors <- function(params, second = FALSE, third = FALSE) {
  # Define for first and total only
  parameters <- rep(params, each = 2)
  sensitivity <- rep(c("Si", "STi"), times = length(params))
  # Define for second only
  pairs <- utils::combn(params, 2, simplify = FALSE)
  parameters.second <- lapply(pairs, function(x) paste0(x, collapse = ".")) %>%
    unlist()
  sensitivity.second <- rep("Sij", times = (length(params) * (length(params) - 1) / 2))
  # Define for third only
  if(length(params) > 2) {
    triplet <- utils::combn(params, 3, simplify = FALSE)
    parameters.third <- lapply(triplet, function(x) paste0(x, collapse = ".")) %>%
      unlist()
    sensitivity.third <- rep("Sijk", times = factorial(length(params)) /
                               (factorial(3) * factorial((length(params) - 3))))
  }
  out <- data.table::as.data.table(cbind(parameters, sensitivity))
  # Set conditions
  if(second == TRUE & third == FALSE) {
    parameters <- c(parameters, parameters.second)
    sensitivity <- c(sensitivity, sensitivity.second)
    out <- data.table::as.data.table(cbind(parameters, sensitivity))
  }
  if(second == FALSE & third == TRUE) {
    parameters <- c(parameters, parameters.third)
    sensitivity <- c(sensitivity, sensitivity.third)
    out <- data.table::as.data.table(cbind(parameters, sensitivity))
  }
  if(second == TRUE & third == TRUE) {
    parameters <- c(parameters, parameters.second, parameters.third)
    sensitivity <- c(sensitivity, sensitivity.second, sensitivity.third)
    out <- data.table::as.data.table(cbind(parameters, sensitivity))
  }
  return(out)
}

#' Bootstrap confidence intervals for Sobol' indices.
#'
#' It computes bootstrap confidence intervals for Sobol' indices.
#'
#' @param b The output of the \code{sobol_indices} function.
#' @param params A vector with the name of the model inputs.
#' @param type A vector of character strings representing
#' the type of intervals required. The value should be any subset
#' of the values \code{c("norm", "basic", "perc", "bca")}. For more information,
#' check the function \code{\link{boot.ci}}.
#' @param conf A scalar or vector containing the confidence
#' level(s) of the required interval(s).
#' @param second Logical. If \code{second = TRUE}, it computes the confidence
#' intervals for second-order indices. Default is \code{second = FALSE}.
#' @param third Logical. If \code{third = TRUE}, it computes the confidence
#' intervals for third-order indices. Default is \code{third = FALSE}.
#'
#' @return A data table.
#' @seealso \code{\link{boot}}, \code{\link{boot.ci}}.
#'
#' @export
#'
#' @examples
#' # Define settings:
#' n <- 1000; k <- 8; R <- 100
#' # Design the sample matrix:
#' A <- sobol_matrices(n = n, k = k, second = TRUE, third = TRUE)
#' # Compute the model output:
#' Y <- sobol_Fun(A)
#' # Compute the Sobol' indices:
#' \donttest{sens <- sobol_indices(Y = Y, params = colnames(data.frame(A)),
#' R = R, n = n, parallel = "no", ncpus = 1,
#' second = TRUE, third = TRUE)
#' # Compute confidence intervals:
#' sobol_ci(sens, params = colnames(data.frame(A)), type = "norm", conf = 0.95)}
sobol_ci <- function(b, params, type, conf, second = FALSE, third = FALSE) {
  .Deprecated(msg = "sobol_ci will be removed in the next version as the computation of confidence intervals will be done directly by sobol_indices")
  out <- sobol_ci_temp(b = b, params = params, type = type,
                        conf = conf, second = second, third = third) %>%
    cbind(create_vectors(params = params, second = second, third = third))
  return(out)
}


# FUNCTION TO COMPUTE CONFIDENCE INTERVALS FOR DUMMY PARAMETER ----------------


#' Bootstrap confidence intervals for the dummy parameter
#'
#' It computes bootstrap confidence intervals for the dummy parameter.
#'
#' @param b The output of the \code{sobol_dummy} function.
#' @param type A vector of character strings representing
#' the type of intervals required. The value should be any subset
#' of the values \code{c("norm", "basic", "perc", "bca")}. For more information,
#' check the function \code{\link{boot.ci}}.
#' @param conf A scalar or vector containing the confidence
#' level(s) of the required interval(s).
#' @importFrom rlang ":="
#'
#' @return A data table.
#' @seealso \code{\link{boot}}, \code{\link{boot.ci}}.
#' @export
#'
#' @examples
#' # Define settings:
#' n <- 100; k <- 8; R <- 10
#' # Design the sample matrix:
#' A <- sobol_matrices(n = n, k = k, second = TRUE, third = TRUE)
#' # Compute the model output:
#' Y <- sobol_Fun(A)
#' # Compute the Sobol' indices for the dummy parameter:
#' s.dummy <- sobol_dummy(Y = Y, params = colnames(data.frame(A)), R = R, n = n)
#' # Compute the confidence intervals for the dummy parameter:
#' sobol_ci_dummy(s.dummy, type = "norm", conf = 0.95)
sobol_ci_dummy <- function(b, type = type, conf = conf) {
  .Deprecated(msg = "sobol_ci_dummy will be removed in the next version as the computation of confidence intervals will be done directly by the function sobol_dummy")
  V1 <- NULL
  sensitivity <- c("Si", "STi")
  parameters <- "dummy"
  out <- lapply(b[, V1], function(x) bootstats(x,
                                               type = type,
                                               conf = 0.95)) %>%
    data.table::rbindlist()
  return(cbind(out, parameters, sensitivity))
}
